runtime/compiler/trj9/runtime/J9Profiler.cpp

```diff
@@ -2448,3 +2507,95 @@ void TR_PersistentProfileInfo::dumpInfo(TR::FILE *logFile)
    if (_valueProfileInfo)
       _valueProfileInfo->dumpInfo(logFile);
    }
+
+TR_AccessedProfileInfo::TR_AccessedProfileInfo(TR::Region &region) :
+    _usedInfo((InfoMapComparator()), (InfoMapAllocator(region))),
+    _current(NULL),
+    _searched(false)
+   {
+   }
+
+TR_AccessedProfileInfo::~TR_AccessedProfileInfo()
+   {
+   // Decrement reference counts for all accessed information
+   for (auto it = _usedInfo.begin(); it != _usedInfo.end(); ++it)
+      {
+      if (it->second)
+        TR_PersistentProfileInfo::decRefCount(it->second);
+      }
+   if (_current)
+      TR_PersistentProfileInfo::decRefCount(_current);
+   }
+
+/**
+ * Compare two sources of profiling information, returning the preferable
+ * one and updating the vmMethod->best if necessary.
+ */
+TR_PersistentProfileInfo *TR_AccessedProfileInfo::compare(TR_PersistentMethodInfo *methodInfo)
+   {
+   if (!methodInfo)
+      return NULL;
+
+   TR_PersistentProfileInfo *recent = methodInfo->getRecentProfileInfo();
+   TR_PersistentProfileInfo *best = methodInfo->getBestProfileInfo();
+
+   // Currently heuristic will replace if recent is non-null
+   // Future implementation should use the method entry block counter to determine
+   // whether enough data has been collected
+   bool replace = recent != NULL && recent != best;
+
+   if (replace)
+      {
+      if (TR::Options::getVerboseOption(TR_VerboseProfiling))
+         TR_VerboseLog::writeLineLocked(TR_Vlog_PROFILING, "For MethodInfo 0x%p, updating best from 0x%p to 0x%p", methodInfo, best, recent);
+      methodInfo->setBestProfileInfo(recent);
+      if (best)
+         TR_PersistentProfileInfo::decRefCount(best);
+      return recent;
+      }
+   else
+      {
+      if (recent)
+         TR_PersistentProfileInfo::decRefCount(recent);
+      return best;
+      }
+   }
+
+/**
+ * For the current method.
+ * More frequenct request, so avoid the overhead of the map.
+ */
+TR_PersistentProfileInfo *TR_AccessedProfileInfo::get(TR::Compilation *comp)
+   {
+   if (_searched)
+      return _current;
+
+   TR::Recompilation *recompilation = comp->getRecompilationInfo();
+   if (recompilation)
+      {
+      TR_PersistentMethodInfo *methodInfo = recompilation->getMethodInfo();
+      _current = compare(methodInfo);
+      if (_current && _current == TR_PersistentProfileInfo::getCurrent(comp))
+         {
+         TR_PersistentProfileInfo::decRefCount(_current); 
+         _current = NULL;
```

Comment:
Does this mean that the only available profiling information is from the current compile, and, since it hasn't executed anything (because it's still being compiled), really there is no profiling information?

Reply:
Yes, there is no point inspecting this information and it may mislead uses of profiling data.